### ISSUE TITLE
New Rule: (Map|Set|Array).forEach async Error

### ISSUE THREAD

**Issue by MentalGear** [eslint/eslint#17843]
### Rule details

Error if .forEach is used with async

forEach does not support async, which is not obvious and can let to hard to debug errors where variables are not available or not filled as expected.

Up to now, ESLINT does not give an error on `Array.forEach(async function(item) {})`

### Related ECMAScript feature

(Map|Set|Array).forEach

### What type of rule is this?

Suggests an alternate way of doing something

### Example code

```js
const myArray = [1, 2, 3, 4, 5];

const newList = []

// This will not work as expected
myArray.forEach(async (item) => {
const value = await doSomething(item)
  newList.push( value )
});

console.log('newList', newList);

// Function to simulate asynchronous operation
function doSomething(item) {
  return new Promise((resolve) => {
    setTimeout(() => {
      return item
      resolve();
    }, 1000);
  });
}

```

### Why should this rule be in the core instead of a plugin?

Native ECMAScript function

### Participation

- [ ] I am willing to submit a pull request to implement this rule.

**Comment by mdjermanovic** [comment]
Hi @MentalGear, thanks for the issue!

This proposal was already evaluated in https://github.com/eslint/eslint/issues/16330 and we decided not to accept it for the reasons described in https://github.com/eslint/eslint/issues/16330#issuecomment-1254048769.

You can configure the [no-restricted-syntax](https://eslint.org/docs/latest/rules/no-restricted-syntax) rule to catch these cases with the following selector:

```
CallExpression[callee.property.name='forEach'] > :matches(ArrowFunctionExpression, FunctionExpression)[async=true].arguments
```

Here's a [playground example](https://eslint.org/play/#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLXJlc3RyaWN0ZWQtc3ludGF4OiBbXCJlcnJvclwiLCB7XG4gICAgc2VsZWN0b3I6IFwiQ2FsbEV4cHJlc3Npb25bY2FsbGVlLnByb3BlcnR5Lm5hbWU9J2ZvckVhY2gnXSA+IDptYXRjaGVzKEFycm93RnVuY3Rpb25FeHByZXNzaW9uLCBGdW5jdGlvbkV4cHJlc3Npb24pW2FzeW5jPXRydWVdLmFyZ3VtZW50c1wiLFxuICAgIG1lc3NhZ2U6IFwiRG8gbm90IHVzZSBhc3luYyBmdW5jdGlvbnMgaW4gLmZvckVhY2goKVwiXG59XSovXG5cbmNvbnN0IG15QXJyYXkgPSBbMSwgMiwgMywgNCwgNV07XG5cbmNvbnN0IG5ld0xpc3QgPSBbXVxuXG4vLyBUaGlzIHdpbGwgbm90IHdvcmsgYXMgZXhwZWN0ZWRcbm15QXJyYXkuZm9yRWFjaChhc3luYyAoaXRlbSkgPT4ge1xuY29uc3QgdmFsdWUgPSBhd2FpdCBkb1NvbWV0aGluZyhpdGVtKVxuICBuZXdMaXN0LnB1c2goIHZhbHVlIClcbn0pO1xuXG5jb25zb2xlLmxvZygnbmV3TGlzdCcsIG5ld0xpc3QpO1xuXG4vLyBGdW5jdGlvbiB0byBzaW11bGF0ZSBhc3luY2hyb25vdXMgb3BlcmF0aW9uXG5mdW5jdGlvbiBkb1NvbWV0aGluZyhpdGVtKSB7XG4gIHJldHVybiBuZXcgUHJvbWlzZSgocmVzb2x2ZSkgPT4ge1xuICAgIHNldFRpbWVvdXQoKCkgPT4ge1xuICAgICAgcmV0dXJuIGl0ZW1cbiAgICAgIHJlc29sdmUoKTtcbiAgICB9LCAxMDAwKTtcbiAgfSk7XG59Iiwib3B0aW9ucyI6eyJlbnYiOnt9LCJydWxlcyI6e30sInBhcnNlck9wdGlvbnMiOnsiZWNtYUZlYXR1cmVzIjp7fSwiZWNtYVZlcnNpb24iOiJsYXRlc3QiLCJzb3VyY2VUeXBlIjoic2NyaXB0In19fQ==) with this selector and code from the original post.
